src/clang.rs

```diff
@@ -1170,6 +1170,157 @@ impl UnsavedFile {
     }
 }
 
+/// Provide index results.
+pub trait IndexCallable {
+    /// Item type for Iterator trait.
+    type Item;
+
+    /// Type reprensenting the number of items.
+    /// If signed, only create iterator if positive.
+    type ItemNum;
+
+    /// Call the function retreiving number of items.
+    fn fetch_item_num(&self) -> Self::ItemNum;
+
+    /// Call the function retreiving the item for the index idx.
+    fn fetch_item(&mut self, idx: usize) -> Self::Item;
+}
+
+/// Tag for unsigned integer type
+pub trait UnsignedInteger {}
+impl UnsignedInteger for c_uint {}
+
+/// Tag for signed integer type
+pub trait SignedInteger {}
+impl SignedInteger for c_int {}
+
+/// An iterator for a type's template arguments
+pub struct IndexCallIterator<CxtT> {
+    cxt: CxtT,
+    length: usize,
+    index: usize
+}
+
+impl<CxtT: IndexCallable> IndexCallIterator<CxtT> where CxtT::ItemNum: UnsignedInteger + Into<u32>{
+    fn new(cxt: CxtT) ->  IndexCallIterator<CxtT> {
+        let len: u32 = cxt.fetch_item_num().into();
+        IndexCallIterator { cxt: cxt, length: len as usize, index: 0 }
+    }
+}
+
+impl<CxtT: IndexCallable> IndexCallIterator<CxtT> where CxtT::ItemNum: SignedInteger + Into<i32> {
+    fn new_check_positive(cxt: CxtT) ->  Option<IndexCallIterator<CxtT>> {
+        let len: i32 = cxt.fetch_item_num().into();
+        if len >= 0 {
+            Some( IndexCallIterator { cxt: cxt, length: len as usize, index: 0 } )
+        } else {
+            debug_assert_eq!(len, -1); // only expect -1 as invalid
+            None
+        }
+    }
+}
+
+impl<CxtT: IndexCallable> Iterator for IndexCallIterator<CxtT> {
+    type Item = CxtT::Item;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.length {
+            let idx = self.index;
+            self.index += 1;
+            Some(self.cxt.fetch_item(idx))
+        } else {
+            None
+        }
+    }
+}
+
+impl<CxtT: IndexCallable> ExactSizeIterator for IndexCallIterator<CxtT> {
+    fn len(&self) -> usize {
+        assert!(self.index <= self.length);
+        self.length - self.index
+    }
+}
+
+struct TestIndexCallableProvider {
```

Comment:
All the testing stuff should be contained in a sub-module at the bottom of this file, like this:

```rust
#[cfg(test)]
mod tests {
    // testing stuff here...
}
```